relay-server/src/actors/store.rs

```diff
@@ -58,33 +58,127 @@ fn make_distinct_id(s: &str) -> Uuid {
+                topic_name: topic_name.to_string(),
+                producer,
+            })
+        }
+        KafkaConfig::Sharded { shards, configs } => {
+            let mut producers = BTreeMap::new();
+            for (shard, kafka_params) in configs {
+                if let Some(producer) = reused_producers.get(&kafka_params.config_name) {
+                    let cached_producer = Arc::clone(producer);
+                    producers.insert(
+                        shard,
+                        (kafka_params.topic_name.to_string(), cached_producer),
+                    );
+                    continue;
+                }
+                for config_p in kafka_params.params {
+                    client_config.set(config_p.name.as_str(), config_p.value.as_str());
+                }
+                let producer = Arc::new(
+                    client_config
+                        .create_with_context(CaptureErrorContext)
+                        .context(ServerErrorKind::KafkaError)?,
+                );
+                reused_producers.insert(kafka_params.config_name, Arc::clone(&producer));
+                producers.insert(shard, (kafka_params.topic_name.to_string(), producer));
+            }
+            Ok(Producer::Sharded { shards, producers })
+        }
     }
+}
 
-    let producer = Arc::new(
-        client_config
-            .create_with_context(CaptureErrorContext)
-            .context(ServerErrorKind::KafkaError)?,
-    );
+/// This object containes the Kafka producer variants for single and sharded configurations.
+pub enum Producer {
+    Single {
+        topic_name: String,
+        producer: Arc<ThreadedProducer>,
+    },
+    Sharded {
+        shards: u64,
+        producers: BTreeMap<u64, (String, Arc<ThreadedProducer>)>,
+    },
+}
 
-    reused_producers.insert(config_name, Arc::clone(&producer));
-    Ok(producer)
+impl Producer {
+    /// Sends the message to Kafka using correct producer for the current topic.
+    fn send(&self, organization_id: u64, message: KafkaMessage) -> Result<(), StoreError> {
+        let serialized = message.serialize()?;
+        metric!(
+            histogram(RelayHistograms::KafkaMessageSize) = serialized.len() as u64,
+            variant = message.variant()
+        );
+        let key = message.key();
+
+        match self {
+            Self::Single {
+                topic_name,
+                producer,
+            } => {
+                let record = BaseRecord::to(topic_name).key(&key).payload(&serialized);
+                producer
+                    .send(record)
+                    .map_err(|(kafka_error, _message)| StoreError::SendFailed(kafka_error))?;
+            }
+
+            Self::Sharded { shards, producers } => {
+                let mut hasher = FnvHasher::default();
+                std::hash::Hash::hash(&organization_id, &mut hasher);
+                let org_id_hash = hasher.finish();
+                let shard = org_id_hash % shards;
```

Comment:
this code exists in two places already, given the nature of it - i suspect both would need updating together all the time - it might be better to create a method somewhere.  Maybe also on the newtype around the BTreeMap I already suggested.

Reply:
moved this into the impl for `Sharded` producer in  [d4a05c2](https://github.com/getsentry/relay/pull/1454/commits/d4a05c23f39c9238481d9338b0b6b7f73ebb7010)